readme.md
# http-request-proxy
It's a package aimed at proxy http request in expressjs. Using this package, you can setup a proxy server easily.

[![build status][action-image]][action-url]
[![Test coverage][coveralls-image]][coveralls-url]
[![David deps][david-image]][david-url]
[![node version][node-image]][node-url]

[npm-url]: https://npmjs.org/package/http-request-proxy
[action-image]: https://github.com/yunnysunny/http-request-proxy/workflows/ci-test/badge.svg
[action-url]: https://github.com/yunnysunny/http-request-proxy/actions?query=workflow%3Aci-test
[coveralls-image]: https://img.shields.io/coveralls/yunnysunny/http-request-proxy.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/yunnysunny/http-request-proxy?branch=master
[david-image]: https://img.shields.io/david/yunnysunny/http-request-proxy.svg?style=flat-square
[david-url]: https://david-dm.org/yunnysunny/http-request-proxy
[node-image]: https://img.shields.io/badge/node.js-%3E=_10-green.svg?style=flat-square
[node-url]: http://nodejs.org/download/

[![NPM](https://nodei.co/npm/node-http-request-proxy.png?downloads=true)](https://nodei.co/npm/node-http-request-proxy/) 

## Install

```npm install http-request-proxy --save```
or
```yarn add http-request-proxy```

## How to use

### Without body-parser

You treat you server as an transparent proxy in this mode, it doesn't modify any HTTP data (both in headers and body).

```javascript
const express = require('express');
const path = require('path');

const {
    TIMEOUT_PROXY,
} = require('./config');
const requestRecordFilter = require('@yunnysunny/request-logging');
const proxy = require('http-request-proxy');

const app = express();
app.enable('trust proxy');

app.set('port', 3003);

app.use(requestRecordFilter());
app.use(proxy({
    urlsToProxy:{
        '/i/back': 'http://localhost:3004'
    },
    beforeParser: true,
    timeout: TIMEOUT_PROXY
}));
```

In this mode , you should set the parameter `beforeParser` to true. And you can set parameter `timeout` to an appropriate value you wanted in millisecond.

### With body-parser

When you want to modify the request parameter in headers or body, you should use this mode. The `dataPrepare` is an function to modify the request parameters, which merged from querystirng and body. The `headerPrepare` is an function to modify the request headers, if it not given , the default headers configuration is `{'User-Agent': 'http-proxy-server'}`.

```javascript
const express = require('express');
const path = require('path');
const bodyParser = require('body-parser');
const multipart = require('connect-multiparty');

const {
    HEADER_SEQ_NUM
} = require('./config');
const requestRecordFilter = require('@yunnysunny/request-logging');
const afterParserProxy = require('http-request-proxy');
const {doSign} = require('./helpers/auth_helper');

const app = express();
app.enable('trust proxy');

app.set('port', 3003);
app.use(requestRecordFilter());

app.use(bodyParser.json({limit: '1mb'}));
app.use(bodyParser.urlencoded({
    extended: false,
    limit: '1mb'
}));
app.use( multipart({}));
app.use(afterParserProxy({
    urlsToProxy:{
        '/i/back': 'http://localhost:3004',
    },
    dataPrepare: function(data) {
        data.sign = doSign(data);
        return data;
    },
    headerPrepare: function(req) {
        return {
            [HEADER_SEQ_NUM]: Number(req.get(HEADER_SEQ_NUM)) + 1
        };
    }
}));
```

But when you want to support uploading in this mode,you should add `connect-multiparty` to your dependencies, and write `app.use( multipart({}))` before call `http-request-proxy`.

It also can set the parameter of `timeout` to control how long you can bear before the response return.

### Using an asynchronous function to determine the destination of backend

Some time you want to request a backend server with dynamic address, and you may request to a dispatcher server to get the backend's ip first. For this purpose, http-request-proxy supply a feature of setting the value of `urlsToProxy` to an an asynchronous function to determine the destination of backend.

```javascript
const express = require('express');
const path = require('path');


const {
    slogger,
    TIMEOUT_PROXY,
    NOT_SUPPORT_URL
} = require('./config');
const requestRecordFilter = require('@yunnysunny/request-logging');
const afterParserProxy = require('http-request-proxy');

const app = express();
app.enable('trust proxy');

app.set('port', 3003);

app.use(requestRecordFilter());
app.use(afterParserProxy({
    urlsToProxy:{
        '/i/back': function(req) {
            return new Promise(function(resolve, reject) {
                if (req.path ==='/i/back' + NOT_SUPPORT_URL) {
                    return reject('not support');
                }
                resolve('http://localhost:3004');
            });
        }
    },
    beforeParser: true,
    timeout: TIMEOUT_PROXY
}));
```


### Using custom configuration to override the warpper's

Sometime you may want to set a special configuration to certain backend server, the others use default setting. You can set one value of `urlsToProxy` to an object, the properties of the object will override the wrapper's.

```javascript
const express = require('express');
const path = require('path');


const {
    slogger,
    TIMEOUT_PROXY,
    NOT_SUPPORT_URL
} = require('./config');
const requestRecordFilter = require('@yunnysunny/request-logging');
const afterParserProxy = require('http-request-proxy');

const app = express();
app.enable('trust proxy');

app.set('port', 3003);

app.use(requestRecordFilter());
app.use(afterParserProxy({
    urlsToProxy:{
        '/i/back': {
            url: function(req) {
                return new Promise(function(resolve, reject) {
                    if (req.path ==='/i/back' + NOT_SUPPORT_URL) {
                        return reject('not support');
                    }
                    resolve('http://localhost:3004');
                });
            },
            timeout: TIMEOUT_CUSTOM
        }
    },
    timeout: TIMEOUT_PROXY
}));
```
The timeout of `/i/back` will use `TIMEOUT_CUSTOM`, the others use `TIMEOUT_PROXY`.

## API

See the document of [api](docs/api.md).



## License

[MIT](LICENSE)

changelog.md
# v0.5.0
## Add
1. The property of `urlsToProxy` of `PorxyOptionsInit` can be an Object which properties can override the PorxyOptionsInit's.

# v0.4.0
## Add
1. Add the property of `_res_data` to `http.ServerResponse` when come across proxy request error, such as the request is timeout or the backend is broken.

# v0.3.0
## Add
1. Add the parameter of `agent` to enable `keep-alive` in http request.

# v0.2.0
## Add
1. Add the parameter of `jsonDisabled` to force the request of `application/json` to be transformed into `application/x-www-form-urlencoded`.

# v0.1.1
## Fix

1. Fix the issue of not sending prepare data to backend when use GET method. 

# v0.1.0
1. Project init.

test/mocha/before.js
const back = require('../express/src/bin/back');
before(function(done) {
    back.start(done);
});

test/express/src/config.js
const slogger = require('node-slogger');


slogger.init({
    flushInterval:500,
});
exports.slogger = slogger;

exports.HEADER_SEQ_NUM = 'seq-num';
exports.HEADER_REQ_NUM = 'req-id';
exports.TIMEOUT_PROXY = 1000;
exports.NOT_SUPPORT_URL = '/not-support';


test/express/src/app_back.js
const express = require('express');
const path = require('path');
const bodyParser = require('body-parser');
const routes = require('./routes/index');

const {
    slogger,
} = require('./config');
const requestRecordFilter = require('@yunnysunny/request-logging');
const {filter: requestValidator} = require('validator-param');

const app = express();
app.enable('trust proxy');

app.set('port', 3004);
app.use(requestRecordFilter());

app.use(bodyParser.json({limit: '1mb'}));
app.use(bodyParser.urlencoded({
    extended: false,
    limit: '1mb'
}));


// app.use(modifyParser);
app.use(requestValidator({
    basePath: path.join(__dirname, './validators'),
    urlPrefix: '/i/',
    filenameReplaces: {
        '_': '/'
    },
    filenameSuffix: '_schema.js'
}));

app.use('/', routes);

app.use(express.static(path.join(__dirname, 'public')));

// catch 404 and forward to error handler
app.use(function(req, res, next) {
    const err = new Error('Not Found:' + req.path);
    err.status = 404;
    next(err);
});

// error handlers
app.use(function(err, req, res, next) {
    const status = err.status;
    if (status === 404) {
        return res.status(404).send(err.message || '未知异常');
    }
    res.status(status || 500);
    slogger.error('发现应用未捕获异常', err);
    res.send({
        msg: err.message || '未知异常',
        code: 0xffff
    });
});


module.exports = app;


index.d.ts
declare function  HttpRequestProxy(
    options: HttpRequestProxy.ProxyOptions,
  ): HttpRequestProxy.NextHandleFunction
  
  declare namespace HttpRequestProxy {
    type NextHandleFunction = (req: any, res: any, next: (err?) => any) => any
    type ProxyUrl = string | Function;
    /**
     * @param {Object} data The original request data, merged by querystring and body.
     * @param {http.IncomingMessage} req The http request object.
     * @returns {Object} The data after prepared.
     */
    type DataPrepareFunction = (data: object, req: http.IncomingMessage) => object
    /**
     * @param {string} proxyUrl The backend url you proxied to 
     * @param {http.ServerResponse} res The express response object
     */
    type ProxyOnErrorCallback = (proxyUrl: string, res: http.ServerResponse) => undefined

    /**
     * @param {http.IncomingMessage} req The express request object.
     * @returns {Object} The headers after prepared.
     */
    type HeaderPrepareFunction = (req: http.IncomingMessage) => object

    interface CommonOptions {
      onError?: ProxyOnErrorCallback;
      dataPrepare?: DataPrepareFunction;
      headerPrepare?: HeaderPrepareFunction;
      /**
       * default: false
       */
      beforeParser?: boolean;
      /**
       * default: false
       */
      jsonDisabled?: boolean;
      /**
       * default: 10000ms
       */
      timeout?: Number;
    }
    interface CustomOptions extends CommonOptions {
      url: ProxyUrl;
    }
    interface UrlsToProxy {
      [key: string]: string | Function | CustomOptions;
    }
    interface ProxyOptions extends CommonOptions {
      urlsToProxy: UrlsToProxy;
    }
  }
  
  export = HttpRequestProxy;

lib/pre_handle.js
const slogger = require('node-slogger');
const {OPTION_KEYS, makeOption} = require('./proxy_option');
const utils = require('./utils');
/**
 * Transform options to the map of path prefix to [CustomOptions](#CustomOptions).
 * 
 * @param { PorxyOptionsInit } options
 * @returns {{urlKeys: Array<String>, urlsToProxy: Object.<String, CustomOptions>} }
 */
module.exports = function preHandle(options) {
    const {urlsToProxy, ...commonOptions} = options;
    const urlKeys = Object.keys(urlsToProxy);
    // transform the option to the map of path prefix
    const allProxyOption = urlKeys.reduce(
        function(options, path) {
            const config = urlsToProxy[path];
            const configValueType = utils.typeof(config);
            switch(configValueType) {
            // common processor
            case 'String': 
            case 'Function':
            case 'AsyncFunction':
                options[path] = makeOption(commonOptions, {url: config});
                break;
                // special processor
            case 'Object': {
                const type = utils.typeof(config.url);
                // not supported type
                if (!['String', 'Function', 'AsyncFunction'].includes(type)) {
                    slogger.error(`The urlsToProxy[${path}] is none of supported type ('String', 'Function', 'AsyncFunction')`);
                    break;
                }
                // init special path option
                const singleOption = makeOption(commonOptions, {url: config.url});
                for (const optionKey of OPTION_KEYS) {
                    if (utils.typeof(config[optionKey]) !== 'Undefined') {
                        singleOption[optionKey] = config[optionKey];
                    }
                }
                options[path] = singleOption;
                break;
            }
            default:
                slogger.error('Not supported config value', path, config);
                break;
            }
            return options;
        }, {},
    );
    return { urlKeys, allProxyOption };
};

lib/proxy_option.js
/** All default option key */
const OPTION_KEYS = [
    'urlsToProxy',
    'onError',
    'dataPrepare',
    'headerPrepare',
    'beforeRequest',
    'afterRequest',
    'beforeParser',
    'jsonDisabled',
    'timeout',
    'agent'
];
/** The part of option default */
const DEFAULT_OPTION = {
    // eslint-disable-next-line no-unused-vars
    onError: function(err, agentUrl) {},
    dataPrepare: function(data) {return data; },
    // eslint-disable-next-line no-unused-vars
    headerPrepare: function(req) { return {}; },
    beforeRequest: function() {},
    afterRequest: function() {},
    beforeParser: false,
    jsonDisabled: false,
    timeout: 10000,
};

/**
 * To create new option Object to avoid the reference coverage by using the args
 * @param  {...any} args 
 */
function makeOption(...args) {
    return Object.assign({}, DEFAULT_OPTION, ...args);
}
module.exports = {
    DEFAULT_OPTION,
    OPTION_KEYS,
    makeOption,
};

lib/proxy.js
const slogger = require('node-slogger');
const request = require('request');
const http = require('http');
const https = require('https');
const fs = require('fs');
const preHandle = require('./pre_handle');
// const {checkUrl, optionProcess} = require('./util');

/**
 * @function DataPrepareFunction
 * 
 * @param {Object} data The original request data, merged by querystring and body.
 * @param {http.IncomingMessage} req The http request object.
 * @returns {Object} The data after prepared.
 */

/**
  * @callback ProxyOnErrorCallback
  * 
  * @param {String} proxyUrl The backend url you proxied to 
  * @param {http.ServerResponse} res The express response object
  */

/**
 * @function HeaderPrepareFunction
 * 
 * @param {http.IncomingMessage} req The express request object.
 * @returns {Object} The headers after prepared.
 */

/**
 * @function ProxyUrlProcess
 * 
 * @param {http.IncomingMessage} req
 * 
 * @returns {Promise}
 */

/** 
  * @typedef {Object} CommonOptions
  * 
  * @property {ProxyOnErrorCallback} onError
  * @property {DataPrepareFunction} dataPrepare
  * @property {HeaderPrepareFunction} headerPrepare
  * @property {Boolean=} [beforeParser=false]
  * @property {Number} [timeout=10000]
  * @property {Boolean=} jsonDisabled When it set true, the type of application/json request will be transformed into application/x-www-form-urlencoded
  * @property {http.Agent=} agent The instance of the class of http.Agent. 
  */
/**
  * @typedef {Object} ProxyUrl 
  * @property {String|ProxyUrlProcess} url
  *
  */
/**
 * @typedef { CommonOptions} CustomOptions
 * 
 * @property {String|ProxyUrlProcess} url
 */
/**
 * @typedef {Object} UrlsToProxy
 * @property {Object.<String, String | ProxyUrlProcess | CustomOptions>} urlsToProxy
 */

/**
 * @typedef {CommonOptions} PorxyOptionsInit
 *
 * @property {Object.<String, String | ProxyUrlProcess | CustomOptions>} urlsToProxy
 */


const checkUrl = function(req, urlsToProxy, urlKeys, callback) {
    const path = req.path;
    const prefix = urlKeys.filter(agentUrl => path.startsWith(agentUrl))[0];
    if (!prefix) {
        return callback(null);
    }
    const proxyOption = urlsToProxy[prefix];
    const needAgentHost = proxyOption.url;
    if (typeof (needAgentHost) === 'string') {
        return  callback(null,needAgentHost + req.originalUrl.slice(prefix.length), proxyOption);
    }
    if (typeof (needAgentHost) === 'function') {
        return needAgentHost(req).then(function(realHost) {
            callback(null,realHost + req.originalUrl.slice(prefix.length), proxyOption);
        }).catch(function(err) {
            callback(err);
        });
    }
    return callback(new Error('not support proxy config ' + path));
};


const _sendStatus = function(res, statusCode, message) {
    res._res_code = message;
    res.status(statusCode).end(message);
};

const _cleanTmpFile = function(files) {
    const keys = Object.keys(files);
    const length = keys.length;
    for(let i=0; i<length; i++) {
        const file = files[keys[i]];
        fs.unlink(file.path, function(err) {
            if (err) {
                slogger.error('delete temporary file failed', err);
            }
        });
    }  
};

const _doProxy = function(options, agentUrl, req, res) {
    const {query, body, method} = req;
    const {
        dataPrepare, timeout, onError, headerPrepare, jsonDisabled, agent
    } = options;
    const reqOptions = {
        url: agentUrl,
        method,
        headers: {},
        timeout,
        agent
    };
    if (options.beforeParser) {
        const httpObj = agentUrl.startsWith('https://') ? https : http;
        reqOptions.headers = req.headers;
        const proxyReq = httpObj.request(agentUrl, reqOptions, function(proxyRes) {
            res.writeHead(proxyRes.statusCode, proxyRes.headers);
            proxyRes.pipe(res);
        });
        proxyReq.setNoDelay(true);
        proxyReq.setTimeout(timeout, function() {
            proxyReq.abort();
            _sendStatus(res, 504, 'proxy ' + agentUrl +' timeout');
        });
        proxyReq.on('error', function(err) {//TODO check
            slogger.error('proxy ' + agentUrl +' error',err);
            _sendStatus(res, 500, 'proxy ' + agentUrl +' error');
        });

        req.pipe(proxyReq);
        return;
    }

    const headers = Object.assign({
        'User-Agent': 'http-proxy-server'
    }, headerPrepare(req));
    reqOptions.headers = headers;
    const data = {};
    Object.assign(data, query, body);

    const dataToSend = dataPrepare(data, req);
    if(req.files) {
        const files = req.files;
        const keys = Object.keys(files);
        const length = keys.length;
        for(let i=0; i<length; i++) {
            const file = files[keys[i]];
            dataToSend[file.fieldName] = {
                value: fs.createReadStream(file.path),
                options: {
                    filename: file.name
                }
            };
        }  
    }

    if (method === 'OPTIONS'|| method === 'GET' || method === 'HEAD') {
        reqOptions.qs = dataToSend;
    } else if (req.is('urlencoded')) {
        reqOptions.form = dataToSend;
    } else if (req.is('multipart')) {
        reqOptions.formData = dataToSend;
    } else if (req.is('json')) {
        if (jsonDisabled) {
            reqOptions.form = dataToSend;
        } else {
            reqOptions.json = true;
            reqOptions.body = dataToSend;
        }
    } else {
        reqOptions.body = dataToSend;
    }
    
    const apiGateWayReq = request(reqOptions, function() {
        if (req.files) {
            _cleanTmpFile(req.files);
        }
    });
    apiGateWayReq.on('error', function(err) {
        slogger.error('request proxy http interface failed', err, reqOptions);
        if (err.code === 'ESOCKETTIMEDOUT' || err.code === 'ETIMEDOUT' ) { //request timeout
            return _sendStatus(res, 504, 'proxy ' + agentUrl +' timeout');
        }
        onError(err, agentUrl);
    });
    apiGateWayReq.pipe(res);
    
};
/** @module http-request-proxy */
/**
 * The express proxy middleware.
 * 
 * @param { PorxyOptionsInit } options
 */
module.exports = function doProxy(options) {
    const {urlKeys, allProxyOption: urlsToProxy} = preHandle(options);
    return function(req, res, next) {
        if (!urlKeys || !urlKeys.length) {
            return next();
        }
        checkUrl(req, urlsToProxy, urlKeys, function(err, backUrl, proxyOption) {
            if (err) {
                const path = req.path;
                slogger.error('proxy ' + path +' failed:', err);
                return _sendStatus(res, 502, 'proxy ' + path +' failed: ' + err);
            }
            if (!backUrl) {
                return next();
            }
            _doProxy(proxyOption, backUrl, req, res);
        });
    };
};
